zmq_addon.hpp

```diff
@@ -37,6 +37,78 @@
 
 namespace zmq
 {
+
+#ifdef ZMQ_CPP11
+
+/*  Receive a multipart message.
+    
+    Writes the zmq::message_t objects to OutputIterator out.
+    The out iterator must handle an unspecified amount of write,
+    e.g. using std::back_inserter.
+    
+    Returns: the number of messages received or nullopt (on EAGAIN).
+    Throws: if recv throws.
```

Comment:
`recv` can probably only throw on the first message part, so in that case, no parts will have been written to `out`.

However, this throws also when `*out++` throws (e.g. `std::bad_alloc`), in which case we are left with a have-received message, of which one part is lost. Maybe that's fine, but should be specified. I am not sure if it would be better to `recv` the remaining parts as well and drop them, to ensure a more consistent state.

Reply:
Ah yes, good point. In that unlikely event I think the caller should handle it, probably by reconnecting the socket, or maybe terminate in the case of bad_alloc.